Implement fn:random-number-generator

LegionMammal978 opened the pull request:
Some implementation notes:

- The `next()` and `permute()` functions are implemented with a new function kind, `anonymous_closure`. These functions go into a separate map in the `StaticFunctions` registry so they won't show up via direct reference or via `fn:function-lookup()`. Their `name()` returns `None`, since it's logically absent. Like `item_last_optional`, they receive `closure_values[0]` as an implicit last argument, to be supplied by `Interpreter::create_static_closure()`.
- When `fn:random-number-generator` is first called, it creates a `u64` seed from the argument (or from the `current_datetime`) using `ahash`. The crate is designed to randomize its results very aggressively on separate runs, so `RandomState::with_seeds(0, 0, 0, 0)` is used to force it to behave deterministically.
- I was somewhat concerned about different `NaN` payloads creating different hashes, but it looks like `ordered-float` [accounts for this](https://docs.rs/ordered-float/5.0.0/src/ordered_float/lib.rs.html#2039-2048) and canonicalizes the payload as it should.
- [`SplitMix64`](https://docs.rs/rand_xoshiro/0.6.0/rand_xoshiro/struct.SplitMix64.html) from `rand_xoshiro` is used as a lightweight PRNG: its total state is no larger than the `u64` seed. Each RNG object stores a copy of the seed in the `next()` and `permute()` closures. Since `SplitMix64` does not offer any simple way to extract its state after calling `next_u64()`, the `next()` function directly inlines its implementation (a `wrapping_add()` with a magic number).
- If an RNG object is used to `permute()` two different sequences with the same length, it will produce the same ordering for both sequences. (Note that since it is deterministic, it should produce the same ordering twice for the *same* sequence.) I considered mixing a hash of the input `Sequence` into the seed value to avoid this, but there is no `Hash` impl for `Sequence`.
- `xee-interpreter` now depends on `rand ^0.8.5` and `rand_xoshiro ^0.6.0`. These are not the latest versions of the crates, since `ibig ^0.3.6` depends on `rand ^0.8.3`, and I didn't want to create duplicate dependencies.

faassen commented:
This is very impressive work, I'm impressed you extended the way functions are defined to have internal ones. I think that might be useful in other cases. I have one question about the inlined code - since it requires a deep understanding and is a potential maintenance hazard, I'd like that code to be well-documented.

> If an RNG object is used to permute() two different sequences with the same length, it will produce the same ordering for both sequences. (Note that since it is deterministic, it should produce the same ordering twice for the same sequence.) I considered mixing a hash of the input Sequence into the seed value to avoid this, but there is no Hash impl for Sequence.

That's indeed weird behavior.   I'm trying to understand this comment in this context:

```rust
// don't use the seed directly, since rejection sampling can cause
// different seeds in a next() sequence to produce the exact same shuffle.
```

I guess that is not sufficient to prevent this behavior as this operation does not mutate the rng object.

It should be possible to create a `Hash` for `Sequence` (and `Item`). I'm going to create an issue for that.

LegionMammal978 commented:
> That's indeed weird behavior. I'm trying to understand this comment in this context:
> 
> ```rust
> // don't use the seed directly, since rejection sampling can cause
> // different seeds in a next() sequence to produce the exact same shuffle.
> ```
>
> I guess that is not sufficient to prevent this behavior as this operation does not mutate the rng object.

That comment is talking about a slightly different issue, which surprised me at first when I saw one of the tests failing. It was checking that `?next()?permute()` did not produce the same shuffle as `?permute()`. Here's a minimal example of the issue:

```rust
/*
[dependencies]
rand = "0.8.5"
rand_xoshiro = "0.6.0"
*/
use rand::prelude::*;
use rand_xoshiro::SplitMix64;

fn main() {
    // take two copies of the same RNG
    let mut rng1 = SplitMix64::seed_from_u64(3);
    let mut rng2 = rng1.clone();
    // advance one of them
    let _ = rng2.next_u64();
    // use both to shuffle a long array
    let mut vec1 = Vec::from_iter(0..128);
    let mut vec2 = Vec::from_iter(0..128);
    vec1.shuffle(&mut rng1);
    vec2.shuffle(&mut rng2);
    // they're the same shuffle!
    assert_eq!(vec1, vec2);
}
```

How it works is that the RNG produces a raw stream of numbers (`0x3be02bda`, `0x31e3716a`, `0x6663bd00`, ...) from calling `next_u32()` repeatedly. But when the `rand` code turns this stream into a sequence of indices to shuffle the slice with, it uses *rejection sampling* to avoid bias. The idea is, it looks at one of the `next_u32()` values in the stream, checks if it can create an unbiased output value from it, and if it can't, then it tries again with the next value in the stream.

The problem is, calling `next_u64()` on the RNG similarly pops the first value off the stream, which can take it into the same state as if rejection sampling had discarded that first value. So in `rng1`, rejection sampling discards the first value `0x3be02bda`, while in `rng2`, the `next_u64()` pops the first value, and now they'll produce exactly the same stream (and exactly the same shuffle) from the second value onward.

My solution is to just take the value and use it to seed a new RNG. This gives the new RNG a stream entirely separate from the original one, so that the 'second value onward' won't come out identical, regardless of how much rejection sampling occurs.

Note that `rand` v0.9.0 seems to have a new sampling method (rust-random/rand#1287) which doesn't discard any values in the stream, and doesn't appear to have this issue. (I've looked at the `cargo-tree`, and `rand ^0.8.0` is mainly needed as a dev-dependency for `proptest`, as well as an optional dependency for `ibig` that we don't actually use.)

LegionMammal978 commented:
I've added some more comments to elaborate on the intent. I've also added a TODO to `permute()` to mix in a sequence hash. (This is pretty trivial to implement: just create a `RandomState::with_seeds(0, 0, 0, 0)` and XOR the seed with `random_state.hash_one(arg)`.)

faassen commented:
Looks good, i'm merging it!
